File: sindrome_overlay/providers/third_party_emotes.py
"""Third-party Twitch emote resolution (BetterTTV, 7TV, FrankerFaceZ).

The resolver fetches the global and channel emote catalogs once per
connection and lets providers augment parsed messages with image emotes for
matching tokens. Native Twitch emotes always take priority: tokens inside a
native emote's range are never replaced.

All network calls are best-effort with short timeouts; failures leave the map
empty and the overlay keeps rendering text fallbacks.
"""

from __future__ import annotations

import logging
import threading
from collections.abc import Iterable
from typing import Any

import requests

from ..models import ChatEmote, ChatMessage

_FETCH_TIMEOUT = (3.0, 6.0)
_MAX_CODE_LENGTH = 40
_MAX_MATCHES_PER_MESSAGE = 30

_BTTV_GLOBAL_URL = "https://api.betterttv.net/3/cached/emotes/global"
_BTTV_CHANNEL_URL = "https://api.betterttv.net/3/cached/users/twitch/{channel_id}"
_SEVENTV_GLOBAL_URL = "https://7tv.io/v3/emote-sets/global"
_SEVENTV_CHANNEL_URL = "https://7tv.io/v3/users/twitch/{channel_id}"
_FFZ_GLOBAL_URL = "https://api.frankerfacez.com/v1/set/global"
_FFZ_ROOM_URL = "https://api.frankerfacez.com/v1/room/id/{channel_id}"


def _bttv_emote_url(emote_id: str) -> str:
    return f"https://cdn.betterttv.net/emote/{emote_id}/2x"


def _seventv_emote_url(emote_id: str) -> str:
    return f"https://cdn.7tv.app/emote/{emote_id}/2x.webp"


def _ffz_url(value: Any) -> str:
    url = str(value or "").strip()
    if url.startswith("//"):
        url = f"https:{url}"
    return url if url.startswith("https://") else ""


def parse_bttv_payload(payload: Any) -> dict[str, str]:
    """Parses a BTTV global emote list or channel/shared emote payload."""
    codes: dict[str, str] = {}
    items: list[Any] = []
    if isinstance(payload, list):
        items = payload
    elif isinstance(payload, dict):
        for section in ("channelEmotes", "sharedEmotes"):
            section_value = payload.get(section)
            if isinstance(section_value, list):
                items.extend(section_value)
    for item in items:
        if not isinstance(item, dict):
            continue
        code = str(item.get("code") or "").strip()
        emote_id = str(item.get("id") or "").strip()
        if code and emote_id:
            codes[code] = _bttv_emote_url(emote_id)
    return codes


def parse_seventv_payload(payload: Any) -> dict[str, str]:
    """Parses a 7TV emote set object or a user payload with an emote_set."""
    codes: dict[str, str] = {}
    emotes: list[Any] = []
    if isinstance(payload, dict):
        emote_set = payload.get("emote_set") if "emote_set" in payload else payload
        raw = emote_set.get("emotes") if isinstance(emote_set, dict) else None
        if isinstance(raw, list):
            emotes = raw
    for item in emotes:
        if not isinstance(item, dict):
            continue
        code = str(item.get("name") or "").strip()
        emote_id = str(item.get("id") or "").strip()
        if code and emote_id:
            codes[code] = _seventv_emote_url(emote_id)
    return codes


def parse_ffz_payload(payload: Any) -> dict[str, str]:
    """Parses an FFZ global or room payload, honouring the room's active set."""
    codes: dict[str, str] = {}
    if not isinstance(payload, dict):
        return codes
    sets = payload.get("sets")
    if not isinstance(sets, dict):
        return codes
    selected: list[Any] = []
    if "default_sets" in payload:
        for set_id in payload.get("default_sets") or []:
            chosen = sets.get(str(set_id))
            if isinstance(chosen, dict):
                selected.append(chosen)
    else:
        room = payload.get("room")
        active_set = room.get("set") if isinstance(room, dict) else None
        chosen = sets.get(str(active_set)) if active_set is not None else None
        if isinstance(chosen, dict):
            selected.append(chosen)
    for set_payload in selected:
        emoticons = set_payload.get("emoticons")
        if not isinstance(emoticons, list):
            continue
        for item in emoticons:
            if not isinstance(item, dict):
                continue
            code = str(item.get("name") or "").strip()
            urls = item.get("urls")
            url = ""
            if isinstance(urls, dict):
                url = _ffz_url(urls.get("2") or urls.get("1") or urls.get("4"))
            if code and url:
                codes[code] = url
    return codes


def _fetch_json(url: str) -> Any:
    response = requests.get(url, timeout=_FETCH_TIMEOUT)
    if response.status_code >= 400:
        return None
    try:
        return response.json()
    except ValueError:
        return None


def _overlaps(occupied: list[tuple[int, int]], start: int, end: int) -> bool:
    for occupied_start, occupied_end in occupied:
        if start < occupied_end and occupied_start < end:
            return True
        if occupied_start >= end:
            break
    return False


def find_code_matches(
    text: str,
    codes: dict[str, str],
    occupied: Iterable[tuple[int, int]] = (),
) -> list[ChatEmote]:
    """Returns image emotes for tokens matching third-party codes.

    Tokens overlapping a native Twitch emote range are skipped so native
    rendering always wins.
    """
    if not text or not codes:
        return []
    occupied_sorted = sorted(occupied)
    matches: list[ChatEmote] = []
    cursor = 0
    for token in text.split():
        start = text.find(token, cursor)
        if start < 0:
            continue
        end = start + len(token)
        cursor = end
        if len(matches) >= _MAX_MATCHES_PER_MESSAGE:
            break
        url = codes.get(token)
        if not url or _overlaps(occupied_sorted, start, end):
            continue
        matches.append(
            ChatEmote(emote_id=token, start=start, end=end, name=token, image_url=url)
        )
    return matches


class ThirdPartyEmoteResolver:
    """Thread-safe code-to-URL map refreshed once per provider connection."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._codes: dict[str, str] = {}
        self._stopped = threading.Event()
        self.log = logging.getLogger("sindrome_overlay.third-party-emotes")

    def stop(self) -> None:
        self._stopped.set()

    def clear(self) -> None:
        with self._lock:
            self._codes.clear()

    def load(self, channel_id: str) -> None:
        if self._stopped.is_set():
            return
        codes: dict[str, str] = {}
        try:
            payload = _fetch_json(_BTTV_GLOBAL_URL)
            if payload is not None:
                codes.update(parse_bttv_payload(payload))
            payload = _fetch_json(_SEVENTV_GLOBAL_URL)
            if payload is not None:
                codes.update(parse_seventv_payload(payload))
            payload = _fetch_json(_FFZ_GLOBAL_URL)
            if payload is not None:
                codes.update(parse_ffz_payload(payload))
            if channel_id.isdigit():
                payload = _fetch_json(_BTTV_CHANNEL_URL.format(channel_id=channel_id))
                if payload is not None:
                    codes.update(parse_bttv_payload(payload))
                payload = _fetch_json(_SEVENTV_CHANNEL_URL.format(channel_id=channel_id))
                if payload is not None:
                    codes.update(parse_seventv_payload(payload))
                payload = _fetch_json(_FFZ_ROOM_URL.format(channel_id=channel_id))
                if payload is not None:
                    codes.update(parse_ffz_payload(payload))
        except requests.RequestException as exc:
            self.log.debug("Third-party emote fetch failed: %s", exc)
            return
        codes = {
            code: url
            for code, url in codes.items()
            if code and len(code) <= _MAX_CODE_LENGTH and url.startswith("https://")
        }
        with self._lock:
            self._codes = codes
        if codes:
            self.log.info("Loaded %d third-party emotes.", len(codes))

    def augment(self, message: ChatMessage) -> ChatMessage:
        with self._lock:
            codes = dict(self._codes)
        if not codes or not message.text:
            return message
        native = [(emote.start, emote.end) for emote in message.emotes]
        extra = find_code_matches(message.text, codes, native)
        if not extra:
            return message
        return ChatMessage(
            platform=message.platform,
            author=message.author,
            text=message.text,
            author_id=message.author_id,
            timestamp=message.timestamp,
            author_colour=message.author_colour,
            badges=message.badges,
            amount=message.amount,
            message_id=message.message_id,
            kind=message.kind,
            emotes=message.emotes + tuple(extra),
            badge_refs=message.badge_refs,
        )

File: android/src/main/java/com/sindromegames/chatoverlay/providers/ThirdPartyEmotes.java
package com.sindromegames.chatoverlay.providers;

import android.util.Log;

import androidx.annotation.Nullable;

import com.sindromegames.chatoverlay.model.ChatEmote;
import com.sindromegames.chatoverlay.model.ChatMessage;
import com.sindromegames.chatoverlay.net.NetClient;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Third-party Twitch emote catalogs (BetterTTV, 7TV, FrankerFaceZ).
 *
 * The catalog is fetched once per connection on a background thread with
 * short timeouts; failures leave the map empty and messages keep rendering
 * plain text. Native Twitch emotes always win: tokens inside a native emote
 * range are never replaced.
 */
public final class ThirdPartyEmotes {
    private static final String TAG = "ThirdPartyEmotes";
    private static final int MAX_CODE_LENGTH = 40;
    private static final int MAX_MATCHES_PER_MESSAGE = 30;

    private volatile Map<String, String> codes = new LinkedHashMap<>();
    private final AtomicBoolean started = new AtomicBoolean(false);

    /** Fetches the catalogs on a background thread; only the first call runs. */
    public void loadOnce(final String channelId, final NetClient net) {
        if (!started.compareAndSet(false, true)) return;
        final Thread loader = new Thread(() -> {
            Map<String, String> loaded = new LinkedHashMap<>();
            fetchAll(channelId, net, loaded);
            if (!loaded.isEmpty()) {
                Log.i(TAG, "Loaded " + loaded.size() + " third-party emotes");
            }
            codes = loaded;
        }, "twitch-third-party-emotes");
        loader.setDaemon(true);
        loader.start();
    }

    private static void fetchAll(String channelId, NetClient net, Map<String, String> output) {
        safeMerge(output, parseBttvArray(getJsonArray(net, "https://api.betterttv.net/3/cached/emotes/global")));
        safeMerge(output, parseSeventv(getJsonObject(net, "https://7tv.io/v3/emote-sets/global")));
        safeMerge(output, parseFfz(getJsonObject(net, "https://api.frankerfacez.com/v1/set/global")));
        if (channelId != null && !channelId.isEmpty()) {
            safeMerge(output, parseBttv(getJsonObject(net,
                    "https://api.betterttv.net/3/cached/users/twitch/" + channelId)));
            safeMerge(output, parseSeventv(getJsonObject(net,
                    "https://7tv.io/v3/users/twitch/" + channelId)));
            safeMerge(output, parseFfz(getJsonObject(net,
                    "https://api.frankerfacez.com/v1/room/id/" + channelId)));
        }
        output.values().removeIf(url -> !url.startsWith("https://"));
    }

    @Nullable private static JSONObject getJsonObject(NetClient net, String url) {
        String body = getRaw(net, url);
        if (body == null) return null;
        try {
            return new JSONObject(body);
        } catch (Exception ignored) {
            return null;
        }
    }

    @Nullable private static JSONArray getJsonArray(NetClient net, String url) {
        String body = getRaw(net, url);
        if (body == null) return null;
        try {
            return new JSONArray(body);
        } catch (Exception ignored) {
            return null;
        }
    }

    @Nullable private static String getRaw(NetClient net, String url) {
        try {
            NetClient.ResponseData response = net.get(url);
            if (response.code() < 200 || response.code() >= 400) return null;
            return response.body();
        } catch (Exception ignored) {
            return null;
        }
    }

    private static void safeMerge(Map<String, String> output, Map<String, String> parsed) {
        for (Map.Entry<String, String> entry : parsed.entrySet()) {
            String code = entry.getKey();
            if (code.isEmpty() || code.length() > MAX_CODE_LENGTH) continue;
            String url = entry.getValue();
            if (url.startsWith("https://")) output.put(code, url);
        }
    }

    /** Parses a BTTV channel payload (channelEmotes + sharedEmotes). */
    public static Map<String, String> parseBttv(@Nullable JSONObject payload) {
        Map<String, String> output = new LinkedHashMap<>();
        if (payload == null) return output;
        List<Object> items = new ArrayList<>();
        addAll(items, payload.optJSONArray("channelEmotes"));
        addAll(items, payload.optJSONArray("sharedEmotes"));
        collectBttv(items, output);
        return output;
    }

    /** Parses a BTTV bare array payload (the global endpoint returns a list). */
    public static Map<String, String> parseBttvArray(@Nullable JSONArray payload) {
        Map<String, String> output = new LinkedHashMap<>();
        List<Object> items = new ArrayList<>();
        addAll(items, payload);
        collectBttv(items, output);
        return output;
    }

    private static void collectBttv(List<Object> items, Map<String, String> output) {
        for (Object raw : items) {
            if (!(raw instanceof JSONObject item)) continue;
            String code = item.optString("code", "").trim();
            String id = item.optString("id", "").trim();
            if (!code.isEmpty() && !id.isEmpty()) output.put(code, bttvUrl(id));
        }
    }

    private static String bttvUrl(String id) {
        return "https://cdn.betterttv.net/emote/" + id + "/2x";
    }

    private static String sevenTvUrl(String id) {
        return "https://cdn.7tv.app/emote/" + id + "/2x.webp";
    }

    public static Map<String, String> parseSeventv(@Nullable JSONObject payload) {
        Map<String, String> output = new LinkedHashMap<>();
        if (payload == null) return output;
        JSONObject emoteSet = payload.has("emote_set") ? payload.optJSONObject("emote_set") : payload;
        if (emoteSet == null) return output;
        JSONArray emotes = emoteSet.optJSONArray("emotes");
        if (emotes == null) return output;
        for (int index = 0; index < emotes.length(); index++) {
            JSONObject item = emotes.optJSONObject(index);
            if (item == null) continue;
            String code = item.optString("name", "").trim();
            String id = item.optString("id", "").trim();
            if (!code.isEmpty() && !id.isEmpty()) output.put(code, sevenTvUrl(id));
        }
        return output;
    }

    public static Map<String, String> parseFfz(@Nullable JSONObject payload) {
        Map<String, String> output = new LinkedHashMap<>();
        if (payload == null) return output;
        JSONObject sets = payload.optJSONObject("sets");
        if (sets == null) return output;
        List<JSONObject> selected = new ArrayList<>();
        if (payload.has("default_sets")) {
            JSONArray defaults = payload.optJSONArray("default_sets");
            if (defaults != null) {
                for (int index = 0; index < defaults.length(); index++) {
                    JSONObject chosen = sets.optJSONObject(String.valueOf(defaults.opt(index)));
                    if (chosen != null) selected.add(chosen);
                }
            }
        } else {
            JSONObject room = payload.optJSONObject("room");
            Object activeSet = room == null ? null : room.opt("set");
            if (activeSet != null) {
                JSONObject chosen = sets.optJSONObject(String.valueOf(activeSet));
                if (chosen != null) selected.add(chosen);
            }
        }
        for (JSONObject set : selected) {
            JSONArray emoticons = set.optJSONArray("emoticons");
            if (emoticons == null) continue;
            for (int index = 0; index < emoticons.length(); index++) {
                JSONObject item = emoticons.optJSONObject(index);
                if (item == null) continue;
                String code = item.optString("name", "").trim();
                JSONObject urls = item.optJSONObject("urls");
                String url = "";
                if (urls != null) {
                    url = normalizeFfzUrl(urls.optString("2", urls.optString("1",
                            urls.optString("4", ""))));
                }
                if (!code.isEmpty() && !url.isEmpty()) output.put(code, url);
            }
        }
        return output;
    }

    private static String normalizeFfzUrl(String value) {
        String url = value == null ? "" : value.trim();
        if (url.startsWith("//")) url = "https:" + url;
        return url;
    }

    private static void addAll(List<Object> target, @Nullable JSONArray array) {
        if (array == null) return;
        for (int index = 0; index < array.length(); index++) {
            Object value = array.opt(index);
            if (value != null) target.add(value);
        }
    }

    /**
     * Returns image emotes for message tokens that match a third-party code
     * and do not overlap a native Twitch emote range.
     */
    public static List<ChatEmote> findMatches(
            String text, Map<String, String> codeMap, List<ChatEmote> nativeEmotes) {
        List<ChatEmote> matches = new ArrayList<>();
        if (text == null || text.isEmpty() || codeMap.isEmpty()) return matches;
        int cursor = 0;
        for (String token : text.split("\\s+")) {
            if (matches.size() >= MAX_MATCHES_PER_MESSAGE) break;
            if (token.isEmpty()) continue;
            int start = text.indexOf(token, cursor);
            if (start < 0) continue;
            int end = start + token.length();
            cursor = end;
            String url = codeMap.get(token);
            if (url == null) continue;
            boolean occupied = false;
            for (ChatEmote nativeEmote : nativeEmotes) {
                if (start < nativeEmote.end && nativeEmote.start < end) {
                    occupied = true;
                    break;
                }
            }
            if (occupied) continue;
            matches.add(new ChatEmote(token, start, end, token, url));
        }
        return matches;
    }

    /** Test-only: injects a catalog without network access. */
    void forceCatalogForTest(Map<String, String> catalog) {
        codes = new LinkedHashMap<>(catalog);
    }

    /** Returns a copy of the message with third-party emotes appended. */
    public ChatMessage augment(ChatMessage message) {
        Map<String, String> current = codes;
        if (current.isEmpty() || message == null || message.text.isEmpty()) return message;
        List<ChatEmote> extra = findMatches(message.text, current, message.emotes);
        if (extra.isEmpty()) return message;
        List<ChatEmote> merged = new ArrayList<>(message.emotes);
        merged.addAll(extra);
        return ChatMessage.builder(message.platform, message.author, message.text)
                .authorId(message.authorId)
                .authorColor(message.authorColor)
                .messageId(message.messageId)
                .amount(message.amount)
                .kind(message.kind)
                .timestamp(message.timestamp)
                .badges(message.badges)
                .emotes(merged)
                .build();
    }
}
